typing/ctype.ml

```diff
@@ -5038,69 +5038,103 @@ let rec arity ty =
   | _ -> 0
 
 (* Check for non-generalizable type variables *)
-exception Nongen
-let visited = ref TypeSet.empty
-
-let rec nongen_schema_rec env ty =
-  if TypeSet.mem ty !visited then () else begin
-    visited := TypeSet.add ty !visited;
+let rec nongen_vars_in_schema_fold env ((visited, acc) as acc') ty =
+  if TypeSet.mem ty visited
+  then acc'
+  else begin
+    let visited = TypeSet.add ty visited in
     match get_desc ty with
-      Tvar _ when get_level ty <> generic_level ->
-        raise Nongen
+    | Tvar _ when get_level ty <> generic_level ->
+        visited, TypeSet.add ty acc
     | Tconstr _ ->
-        let old = !visited in
-        begin try iter_type_expr (nongen_schema_rec env) ty
-        with Nongen -> try
-          visited := old;
-          nongen_schema_rec env (try_expand_head try_expand_safe env ty)
-        with Cannot_expand ->
-          raise Nongen
+        let (_, unexpanded_result) as unexpanded_result' =
+          fold_type_expr
+            (nongen_vars_in_schema_fold env)
+            (visited, acc)
+            ty
+        in
+        if unexpanded_result == acc
+        then (visited, acc)
+        else begin
+          match
+            nongen_vars_in_schema_fold env (visited, acc)
+              (try_expand_head try_expand_safe env ty)
+          with
+          | exception Cannot_expand -> unexpanded_result'
+          | expanded_result -> expanded_result
         end
     | Tfield(_, kind, t1, t2) ->
-        if field_kind_repr kind = Fpublic then
-          nongen_schema_rec env t1;
-        nongen_schema_rec env t2
+        let visited, acc =
+          match field_kind_repr kind with
+          | Fpublic -> nongen_vars_in_schema_fold env (visited, acc) t1
+          | _ -> visited, acc
+        in
+        nongen_vars_in_schema_fold env (visited, acc) t2
     | Tvariant row ->
-        iter_row (nongen_schema_rec env) row;
-        if not (static_row row) then nongen_schema_rec env (row_more row)
+        let visited, acc =
+          fold_row (nongen_vars_in_schema_fold env) (visited, acc) row
+        in
+        if not (static_row row)
+        then nongen_vars_in_schema_fold env (visited, acc) (row_more row)
+        else (visited, acc)
     | _ ->
-        iter_type_expr (nongen_schema_rec env) ty
+        fold_type_expr (nongen_vars_in_schema_fold env) (visited, acc) ty
   end
 
-(* Return whether all variables of type [ty] are generic. *)
-let nongen_schema env ty =
-  visited := TypeSet.empty;
-  try
-    nongen_schema_rec env ty;
-    visited := TypeSet.empty;
-    false
-  with Nongen ->
-    visited := TypeSet.empty;
-    true
+(* Return all non-generic variables of [ty]. *)
+let nongen_vars_in_schema env ty =
+  let _, result =
+    nongen_vars_in_schema_fold env (TypeSet.empty, TypeSet.empty) ty
+  in
+  if TypeSet.is_empty result
+  then None
+  else Some result
+
+let nongen_vars_in_schema_fold_set env acc ty =
```

Comment:
I'm not happy with the naming of these utility functions, but they make `nongen_class_type` much easier to read.